jcl/src/openj9.dtfj/share/classes/com/ibm/dtfj/javacore/parser/j9/section/threadhistory/ThreadHistorySectionParser.java

```diff
@@ -0,0 +1,153 @@
+import com.ibm.dtfj.javacore.parser.j9.IAttributeValueMap;
+import com.ibm.dtfj.javacore.parser.j9.SectionParser;
+import com.ibm.dtfj.javacore.parser.j9.section.common.TraceTimeParser;
+
+/**
+ * Provides parser for Thread history section in the javacore.
+ */
+public class ThreadHistorySectionParser extends SectionParser implements IThreadHistoryTypes {
+
+	private IImageAddressSpaceBuilder fImageAddressSpaceBuilder;
+	private IImageProcessBuilder fImageProcessBuilder;
+	private IJavaRuntimeBuilder fJavaRuntimeBuilder;
+
+	/**
+	 * Construct a javacore parser for the trace
+	 * of the current thread.
+	 */
+	public ThreadHistorySectionParser() {
+		super(THREADHISTORY_SECTION);
+	}
+
+	/**
+	 * Overall controls of parsing for the thread history section.
+	 * @throws ParserException on failure to parse the history due to IOException or excessive line length
+	 */
+	@Override
+	protected void topLevelRule() throws ParserException {
+		fImageAddressSpaceBuilder = fImageBuilder.getCurrentAddressSpaceBuilder();
+		fImageProcessBuilder = fImageAddressSpaceBuilder.getCurrentImageProcessBuilder();
+		fJavaRuntimeBuilder = fImageProcessBuilder.getCurrentJavaRuntimeBuilder();
+
+		parseThreadHistoryLine();
+	}
+
+	/**
+	 * Parse the thread history information.
+	 * @throws ParserException on failure to parse the history due to IOException or excessive line length
+	 */
+	private void parseThreadHistoryLine() throws ParserException {
+		IAttributeValueMap results;
+
+		// Process the history lines
+		while ((results = processTagLineOptional(T_1XECTHTYPE)) != null) {
+			String name = results.getTokenValue(THREADHISTORY_NAME);
+			List<String> lines = new ArrayList<>();
+			boolean first = true;
+			IAttributeValueMap resultsTraceLine;
+			while ((resultsTraceLine = processTagLineOptional(T_3XEHSTTYPE)) != null) {
+				// 3XEHSTTYPE
+				// 04:36:21:235675300 GMT j9dmp.9 -   Preparing for dump, filename=C:\Users\andre\javacore.20220930.053619.81568.0001.txt
+				String hist = resultsTraceLine.getTokenValue(THREADHISTORY_LINE);
+				lines.add(hist);
+				if (first) {
+					first = false;
+					ImageBuilder imageBuilder = (ImageBuilder)fImageBuilder;
+					try {
+						long oldtime = imageBuilder.getImage().getCreationTime();
+						long newtime = TraceTimeParser.parseTimeAdjust(hist, oldtime, 15);
+						if (oldtime != newtime) {
+							fImageBuilder.setCreationTime(newtime);
+						}
+					} catch (DataUnavailable e) {
+						continue;
+					}
+				}
+			}
+			// lines appear in Javacore as newest first, so reverse
+			Collections.reverse(lines);
+			fJavaRuntimeBuilder.addTraceBuffer(name, lines);
+		}
+
+		/*
+		 * Fix up JVM start time.
+		 * The start time doesn't have a time zone, so is parsed according
+		 * to the parser's timezone, which might differ from the Javacore.
+		 * Once the javacore time is fixed up, the offset between the
+		 * JVM start time and nanos should be the same as the offset between
+		 * the javacore time and nanos.
+		 */
```

Comment:
This doesn't seem like the right place for this.

Reply:
Main comments moved to header of method.